t/client-connect-async.t

```diff
@@ -0,0 +1,130 @@
+use strict;
+use warnings;
+use OPCUA::Open62541 ':all';
+use POSIX qw(sigaction SIGALRM);
+
+use Net::EmptyPort qw(empty_port);
+use Test::More tests => 24;
+use Test::LeakTrace;
+
+# initialize the server
+
+my $s = OPCUA::Open62541::Server->new();
+ok($s, "server");
+
+my $sc = $s->getConfig();
+ok($s, "config server");
+
+my $port = empty_port();
+my $r = $sc->setMinimal($port, "");
+is($r, STATUSCODE_GOOD, "minimal server config");
+
+my $pid = fork // die "Unable to fork: $!\n";
+
+if ( !$pid ) {
+    my $running = 1;
+    sub handler {
+	$running = 0;
+    }
+
+    # Perl signal handler only works between perl statements.
+    # Use the real signal handler to interrupt the OPC UA server.
+    # This is not signal safe, best effort is good enough for a test.
+    my $sigact = POSIX::SigAction->new(\&handler)
+	or die "could not create POSIX::SigAction";
+    sigaction(SIGALRM, $sigact)
+	or die "sigaction failed: $!";
+    alarm(1)
+	// die "alarm failed: $!";
+
+    # run server and stop after one second
+    $s->run($running);
+
+    POSIX::_exit 0;
+}
+
+my @testdesc = (
+    ['client', 'client creation'],
+    ['config', 'config creation'],
+    ['config_default', 'set default config'],
+    ['connect_async', 'call to connect_async'],
+    ['iterate', 'calls to run_iterate'],
+    ['state_session', 'client state SESSION after connect'],
+    ['cb_client', 'client in callback'],
+    ['cb_data', 'data in callback'],
+    ['cb_after', 'callback called after connect'],
+    ['disconnect', 'client disconnected'],
+    ['state_disconnected', 'client state DISCONNECTED after disconnect'],
+);
+my %testok = map { $_ => 0 } map { $_->[0] } @testdesc;
+
+no_leaks_ok {
+    my $c;
+    my $data = ['foo'];
+    {
+	$c = OPCUA::Open62541::Client->new();
+	$testok{client} = 1 if $c;
+
+	my $cc = $c->getConfig();
+	$testok{config} = 1 if $cc;
+
+	$r = $cc->setDefault();
+	$testok{config_default} = 1 if $r == STATUSCODE_GOOD;
+
+	$r = $c->connect_async(
+	    "opc.tcp://localhost:$port",
+	    sub {
+		my ($c, $d) = @_;
+		$testok{cb_client} = 1 if $c->getState == CLIENTSTATE_SESSION;
+		$testok{cb_data} = 1 if $d->[0] eq 'foo';
+		push(@$data, 'bar');
+	    },
+	    $data
+	);
+	$testok{connect_async} = 1 if $r == STATUSCODE_GOOD;
+
+	my $failed_iterate = 0;
+	while($c->getState != CLIENTSTATE_SESSION) {
```

Comment:
I don't like tests that may run forever.  They hang if something goes wrong.  Could you run the loop for e.g. 1000 times and return an error if the CLIENTSTATE_SESSION was not created before.  Or you add a timeout and abort if the loop has not finished in that time.